kirtonBCIlab/boccia-bci
# BOC143 – Centralize Ramp Paramters

[BOC143](https://bci4kids.atlassian.net/browse/BOC-143)

Parameters impacting Ramp and Fan settings have been centralized through the introduction of RampSettings and FanSettings containers within BocciaData and BocciaModel.

Code has been updated throughout to make it work with this.

- Included in this is that the limits of the Game Options UI sliders are pulled from these parameters.
- The origin, minimum and maximum elevation and rotation of the ramp (which propagates to Simulated and Hardware ramp, the clamp functions within them and the reset buttons).
- The FanSettings min/max of elevation and rotation range, and elevation and precision max.

I also fixed a number of scripts that use both Start() and OnEnable() that would throw-up NullRef because OnEnable() tries to access the model before its reference is created in Start(). The solution was to put the model reference in Awake()

# IMPORTANT NOTE
- A behaviour in FanSettings Limits (min/max of elevation or rotation range or precision) is that if the limits are changed such that the "default" settings of the CoarseFanSettings scriptable object are outside these limits, then the CoarseFanSettings scriptable objects values will get clamped to within this range – **and it will not revert if the limits are set again to the previous value**. 
- E.g. if RotationRangeMax (Theta) is changed from 180 to 120, this will clamp Theta for CoarseFanSettings to 120 from the "Default" value of 160 – and it will not revert to the "Default" value of 160 if RotateRangeMax (Theta) is changed back to 180. 
- 
![Screenshot 2024-11-13 at 15 31 01](https://github.com/user-attachments/assets/a5880804-2483-41c9-aa4c-28fd9931c849)
**I have attached a screenshot of the "Default" CoarseFanSettings in case this needs to be set back to the original "Default" values in the course of PR testing**

# Testing
- Go to Lines 165 to 212 within BocciaModel
- Modify various settings (e.g. Elevation or Rotation Origin, or Min/Max Limits). **Important: See note above about FanSettings that may impact the CoarseFanSettings scriptable object**
- See that the settings you changed impact the game, e.g. the rotation or elevation origin, their limits, or settings that also show up in the GameOptionsMenu (e.g. Speed in SetRampSettings() or all of SetFanSettings()
- Revert changes to see that these do revert behaviours
- **Again, see note about CoarseFanSettings**
